### ISSUE TITLE
Class members don't exist at runtime but pyright allows reference to them without an error

### ISSUE THREAD

**Issue by kyle-query** [microsoft/pyright#3886]
**Describe the bug**

After declaring a class that has annotated members, pyright allows you to refer to them as members of that class (even though they only exist on instances of the class).

**To Reproduce**

```python
# test.py:
class T:
    t: int

x = T.t # pyright infers x has type int
```

```
$ python test.py
AttributeError: type object 'T' has no attribute 't'
```

**Expected behavior**

pyright should indicate an error in the expression `T.a`, along the lines of "cannot access member a for Type[T]]. Member 'a' is unknown".

**Screenshots or Code**

See above

**VS Code extension or command-line**

I am running pyright as a language server connected to neovim. The errors also occurs on the command line with pyright version 1.1.267.

**Comment by erictraut** [comment]
This is intended. The problem is that it's legitimate for the class member to be assigned elsewhere, perhaps within another module or a subclass. Pyright is able to detect the use of unassigned local variables, but what you're suggesting would require global program analysis, which isn't feasible or even theoretically possible in many cases because the order of execution cannot be determined statically.

**Comment by kyle-query** [comment]
Can you demonstrate how it could be assigned elsewhere? I know pyright catches this error, so I think I'm missing what you mean:

```python
class T:
    t: int

x = T.t # no error
y = T.u # error: Cannot access member "u" for type "Type[T]". Member "u" is unknown
```

**Comment by erictraut** [comment]
Those are two very different situations even though they might look similar at first glance. In the case of `y = T.u`, you're accessing an attribute of the class that hasn't been declared. That's clearly a type violation, and it's the job of a type checker to detect and report such violations.

In the case of `x = T.t` (in your first code sample above), you're accessing a class variable that has been declared by the class author to be present and have a type of `int`. It's the job of a type checker to make sure that any value assigned to this class variable is compatible with `int`. A type checker can also assume that if you read from this declared class variable, the resulting value will be type `int` (or a subtype thereof).

To answer your question, the value can be assigned anywhere — in another module that gets called, in a subclass, etc. The type checker doesn't necessarily have visibility into all code paths within a program. 

Pyright does offer a super-strict diagnostic rule called `reportUninitializedInstanceVariable`. This rule isn't even enabled in `strict` mode because it imposes a bunch of additional restrictions on how you must write your code. But with those restrictions and assumptions pyright will catch the error you mentioned. Try that and see if it suits your needs.

I'll note that in your last code sample above, you've added a `@dataclass` decorator, which changes the situation pretty significantly from your earlier code sample, which didn't include `@dataclass`. The `dataclass` function modifies a class at runtime and turns its class variables into instance variables. There is a bunch of special-case logic in pyright to deal with this special behavior, so the behavior you see with dataclasses may differ from non-dataclasses.

**Comment by kyle-query** [comment]
Ok, I think your comment helped me narrow where my confusion is.

```python
class T:
    t: int

    # I believe this is functionally equivalent to the constructor @dataclass would generate
    def __init__(self, t: int):
        self.t     # this is read as <class T>.t
        self.t = 1 # this is read as <T object>.t

T.t     # this is read as <class T>t and it's unbound here
T.t = 1 # now it's bound. pyright ensures you only assign an int
T.t     # 1

T(0).t # this is read as <T object>.t, and is 0
```

So it seems `t: int` below `class T:` is read as declaring both a class variable and an instance variable.

I thought t was **a class variable only when** you have `t: ClassVar[int]` or assign a value `t: int = 1` inside the class. Instead, it's **only a class variable when** you have that`t: ClassVar[int]`; it's not also an instance variable.

So to summarize,
* Both a class and instance variable, `class T:   t: int`, each in separate namespaces
* Only a class variable: `class T:    t: ClassVar[int]`
* Only an instance variable:

```python
class T:
    def __init__(self):
        self.t: int

w = T.t # Cannot access member "t" of type "Type[T]". Member "t" is unknown
x = T().t # x has the inferred type int
    # This will cause a runtime error because <T object>.t wasn't initialized, much like the original
    # example in the issue description, except this is at an instance level
```

In example though, `typing.get_type_hints` doesn't return any information in this case (unlike the other two cases). Is there another way to declare instance-only variables so their type hints are visible in `typing.get_type_hints`?

Pyright must be getting hints somehow, because it knows `T().t` has type `int`.

**Comment by kyle-query** [comment]
For reference, here's where I read that the original example could would only declare an instance variable: https://peps.python.org/pep-0526/#class-and-instance-variable-annotations

**Comment by erictraut** [comment]
Yes, I think you've got it. For more details, refer to [this documentation](https://github.com/microsoft/pyright/blob/main/docs/type-concepts.md#class-and-instance-variables).

`typing.get_type_hints` is a runtime call. Pyright doesn't use any runtime facilities. (Pyright isn't written in Python and runs fine even if there is no Python interpreter installed on your machine.) Pyright gets all of its information from statically analyzing the Python code.

Is there a reason you want to use `typing.get_type_hints` to access type annotations at runtime? Or are you just curious about how pyright works?

**Comment by kyle-query** [comment]
Thanks Eric, that's a good reference.

> Is there a reason you want to use typing.get_type_hints to access type annotations at runtime? Or are you just curious about how pyright works?

Yes, some of both.

I've written something to automatically derive coercion functions from JSON-like data according to a type-annotated function or class; it validates that runtime input is compatible with type annotations and produces sensible error messages when it's not possible to convert data safely. This is similar to pydantic, but it works without modification to your classes, and it also works to invoke functions besides constructors.

I'm now building an embedded DSL that uses type-hinted classes as input. Perhaps it won't pan out, but I'd like the user to be able to perform operations on values of these types just as they would in python, but rather than their expression being evaluated, an AST will be built. Imagine the user giving you a `lambda` with type `Callable[[int], T]`, but rather than calling it with an int, you pass it a proxy that has overloaded `__and__`, `__sub__`, etc and each of these builds up the AST. The generated AST will be type-correct by construction because the type checker will catch type errors in the `lambda` when the user writes it; no need to implement a type checker for the AST.

Type theory has been an interest of mine for many years, but I'm only now coming back to Python; type hints have been added since I last used Python, so it's been a profound improvement! I'm very impressed by how successful type checkers have been in dynamic languages. I recall first learning about gradual typing a decade ago (from Sam Tobin-Hochstadt's and Matthias Felleisen's work), and it seemed pretty far away from becoming a practical reality.

**Comment by kyle-query** [comment]
> In this last example though, typing.get_type_hints doesn't return any information in this case (unlike the other two cases). Is there another way to declare instance-only variables so their type hints are visible in typing.get_type_hints?

Wanted to close the loop on this question. This is not possible -- while pyright tracks that information, Python's runtime does not:

```python
import typing

class T:
    t: int
    def __init__(self, u: int):
        self.v: int = 1

>>> typing.get_type_hints(T)
{'t': <class 'int'>}

>>> typing.get_type_hints(T.__init__)
{'u': <class 'int'>}

# v is not visible to get_type_hints at runtime, even though its is known statically by pyright
```

Earlier you said,

> The dataclass function modifies a class at runtime and turns its class variables into instance variables. There is a bunch of special-case logic in pyright to deal with this special behavior, so the behavior you see with dataclasses may differ from non-dataclasses.

This makes sense now. Indeed, dataclasses confused the issue for me a bit. It seems like `@dataclass` also adds type annotations to `__init__` by modifying `T.__init__.__annotations__` at runtime. Not knowing that, I guessed that `get_type_annotations` did some kind of static analysis to infer things that weren't explicitly written. Thanks again for helping me clear up some misunderstandings.
